content/docs/user-guide/data-pipelines/defining-pipelines.md

```diff
@@ -0,0 +1,237 @@
+stages with `dvc stage add` -- a limited command-line interface to setup
+pipelines. Let's add another stage this way and look at the resulting
+`dvc.yaml`:
+
+```dvc
+$ dvc stage add --name train \
+                --deps src/model.py \
+                --deps data/clean.csv \
+                --outs data/predict.dat \
+                python src/model.py data/clean.csv
+```
+
+```yaml
+stages:
+  prepare:
+    ...
+    outs:
+      - data/clean.csv
+  train:
+    cmd: python src/model.py data/model.csv
+    deps:
+      - src/model.py
+      - data/clean.csv
+    outs:
+      - data/predict.dat
+```
+
+<admon type="tip">
+
+One advantage of using `dvc stage add` is that it will verify the validity of
+the arguments provided (otherwise stage definition won't be checked until
+execution). A disadvantage is that some advanced features such as [templating]
+are not available this way.
+
+[command]: /doc/user-guide/project-structure/dvcyaml-files#stage-commands
+[templating]: /doc/user-guide/project-structure/pipelines-files#templating
+
+</admon>
+
+Notice that the new `train` stage depends on the output from stage `prepare`
+(`data/clean.csv`), forming the [DAG](#directed-acyclic-graph-dag).
+
+[tracked by dvc]: /doc/start/data-management
+
+## Simple dependencies
+
+There's more than one type of stage dependency. A simple dependency is a file or
+directory used as input by the stage command. When it's contents have changed,
+DVC "invalidates" the stage -- it knows that it needs to run again. This in turn
+may cause a chain reaction in which subsequent stages of the
+<abbr>pipeline</abbr> are also reproduced.
+
+<admon type="info">
+
+DVC [calculates a hash] of file/dir contents to compare vs. previous versions.
+This is a distinctive mechanism over traditional build tools like `make`.
+
+[calculates a hash]:
+  /doc/user-guide/project-structure/internal-files#structure-of-the-cache-directory
+
+</admon>
+
+File system-level dependencies are defined in the `deps` list of `dvc.yaml`
+stages; Alternatively, using the `--deps` (`-d`) option of `dvc stage add` (see
+the previous section's example).
+
+<details>
+
+### External dependencies: click to learn more.
+
+A less common kind of dependency is a _URL dependency_. Instead of files in a
+local disk, you can `dvc import` data from another <abbr>DVC project</abbr> (for
+example hosted on GitHub). External dependencies establish relationships between
+different projects or systems (see `dvc import-url`).
+[Get all the details](/doc/user-guide/external-dependencies).
+
+<admon type="info">
+
+DVC will use special methods to check whether the contents of an URL have
+changed for the purpose of stage invalidation.
+
+</admon>
+
+</details>
+
+## Parameter dependencies
+
+A more narrow type of dependency is the parameter (`params` list in `dvc.yaml`
+stages), or _hyperparameters_ in machine learning. These represent simple values
```

Comment:
simple - remove? (it can be array actually?)

Reply:
"simple values" is how we define parameters conceptually. The fact that you can use arrays is kind of an usage detail: it just means that you can point to groups of params (which are simple values at the tip of the tree).

Reply:
hmm. not sure I understand :) I would remove it tbh ... your call

Reply:
Moved this discussion to #3899.